### ISSUE TITLE
Adding more parallelism to async

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#1586]
Currently the async uses one timer at the background.

In the future this could be multiple timers or using parallel extensions. This should be configurable.

Writing in parallel to multiple files will boost the performance, but we should also check if this counts for writing to the same file/database etc.

**Comment by 304NotModified** [comment]
More timers or using parallel extensios

**Comment by luigiberrettini** [comment]
You could also make use of TPL and concurrent queues rather than relying on locks.
Attention should be paid to concurrency and the need of atomic operations: look at the AsynWrapperTarget and the missing lock on queue.count operations

**Comment by 304NotModified** [comment]
Thanks for input!



> You could also make use of TPL and concurrent queues rather than relying on locks.

Agree, I meant that also with "parallel extensions." ;)  Haven't worked with those yet, but think that's a good option.


> look at the AsynWrapperTarget and the missing lock on queue.count operations

Could you elaborate on that? Is there a bug in the AsynWrapperTarget ?

**Comment by luigiberrettini** [comment]
Maybe: I do not have the big picture about what it does, but the [count operation on AsyncRequestQueue](https://github.com/NLog/NLog/blob/cdb1fcf14f6e505317ab9cdd00322fe60ba4b4ca/src/NLog/Targets/Wrappers/AsyncRequestQueue-T.cs#L74) is not thread-safe

Moreover [ProcessPendingEvents](https://github.com/NLog/NLog/blob/cdb1fcf14f6e505317ab9cdd00322fe60ba4b4ca/src/NLog/Targets/Wrappers/AsyncTargetWrapper.cs#L284-L296) could have issues.

**Comment by 304NotModified** [comment]
Good one. It's missing a lock indeed.

**Comment by Jeinhaus** [comment]
I think the problem might be, that the wrapped target also has to be threadsafe and has to benefit of multiple threads writing to it. If you take the FileTarget for example, it only benefits from parallel writing, if the messages are written to multiple files. And if it doesn't, it has to make sure, that concurrent threads don't mess anything up.

**Comment by luigiberrettini** [comment]
As for the use of TPL take a look at the approach I used in [the Syslog target](https://github.com/graffen/NLog.Targets.Syslog/pull/45)

**Comment by 304NotModified** [comment]
Cool! Really helpful. 

Could you help with implementing this in the asyncwrapper?

**Comment by luigiberrettini** [comment]
Sure, if you are not in a hurry.
Let me see how it behaves and start a "conversation" about the way we want to evolve the AsyncWrapper

**Comment by 304NotModified** [comment]
cool!

No not in a hurry, almost holiday here, so really busy now and within a week relax mode :)

**Comment by luigiberrettini** [comment]
TPL is not supported in .NET 3.5, Silverlight 4 because it was introduced later.
How would you prefer to approach this?

**Comment by 304NotModified** [comment]
We prefer to implement it it the supported frameworks, by using `#if TLP` and add the debug symbol to the csproj files.

**Comment by luigiberrettini** [comment]
If I understand correctly, unsupported frameworks will keep the current version and supported ones will use the new version by means of preprocessor directives.
I do no understand what you mean with "add the debug symbols": debug info full?

![debug info full](http://i.stack.imgur.com/M3S1x.png)

**Comment by 304NotModified** [comment]
I meant conditional compilation symbols. http://haacked.com/images/haacked_com/WindowsLiveWriter/ConditionalConstantsandASP.NET_12C5D/ConditionalCompilation%20-%20Microsoft%20Visual%20Studio_1.png

The code will be backwards compatible (also in behavior) right? So we only extending the possibilities and only on the platforms where its possible (eg .net4+)

**Comment by luigiberrettini** [comment]
As soon as I solve a performance issue with my code I will begin looking at the AsyncWrapper that will be based on the same concepts.
Sorry but it may take long: profilers are not proving so helpful :disappointed:

**Comment by 304NotModified** [comment]
Ow no problem! This isn't a feature which has high prio, but it would be very cool to have. :)

I prefer small steps with good code rather then a quick fix.

**Comment by snakefoot** [comment]
Writing out-of-order in parallel sounds like lots of unwanted side effects in the final result. Very hard for the async-wrapper to parallelize the writing for an unknown Target. It is pretty much only the dequeuing and the adding of continuation-wrappers.

A little improvement has been made to the async-wrapper, so now it can schedule instant timers, instead of waiting 1 ms (or more) between each write operation. See #1684

**Comment by luigiberrettini** [comment]
Unfortunately I overestimated my free time and cannot manage to find time to implement a TPL version of the AsyncTargetWrapper.
Moreover behavior would change and we would need to implement back-pressure and multiple Tasks processing the enqueued log entries.
I do not agree with @snakefoot since, generally speaking, a log system should not care about message order: time information help you sort messages when debugging, therefore you do not need to have a knowledge of which specific target is being wrapped.

**Comment by snakefoot** [comment]
I have no issue with making the queue-handling more parallel (easy task), but I was reading it as there should be multiple async-threads performing the dequeue and write operation. Then you suddenly need to introduce a whole new Target-interface for supporting multiple threads within the same Target (not easy task).

**Comment by 304NotModified** [comment]
I think it's not required to have the events written in order, as Luigi states. We have the longtime for that. 

My use case would be multi threaded write to a database. 

So I think we don't need a new interface or something like that. 

Any way, this feature has low priority and if you like to work on something more requested ;), then it would be nice to spent some time on filetarget bugs, structural logging (60% now) or netcore support (also 90%)

**Comment by snakefoot** [comment]
Sounds like a multiplexer-target infront of several async-targets that each point to their own database-target, that again point to the same database. The multiplexer-target then just performs round-robin scheduling of the write-operations against the async-targets.

**Comment by snakefoot** [comment]
Looks like the performance issue is actually the bottleneck in the Monitor-lock used by the async-wrapper. This causes congestion when having many concurrent logger-threads writing to the same async-target. See also #2650

**Comment by 304NotModified** [comment]
no plans for this
